File: contests/lc/1114.PrintinOrder/print-in-order.go
package main

import (
	"fmt"
	"sync"
)

type Foo struct {
	firstDone  chan struct{}
	secondDone chan struct{}
}

func Constructor() *Foo {
	return &Foo{
		firstDone:  make(chan struct{}),
		secondDone: make(chan struct{}),
	}
}

func (f *Foo) First(printFirst func()) {
	printFirst()
	close(f.firstDone)
}

func (f *Foo) Second(printSecond func()) {
	<-f.firstDone
	printSecond()
	close(f.secondDone)
}

func (f *Foo) Third(printThird func()) {
	<-f.secondDone
	printThird()
}

func main() {
	foo := Constructor()

	var wg sync.WaitGroup
	wg.Add(3)

	go func() {
		defer wg.Done()
		foo.Third(func() { fmt.Print("third") })
	}()

	go func() {
		defer wg.Done()
		foo.Second(func() { fmt.Print("second") })
	}()

	go func() {
		defer wg.Done()
		foo.First(func() { fmt.Print("first") })
	}()

	wg.Wait()
	fmt.Println()
}

File: contests/lc/1114.PrintinOrder/print_in_order.py
import threading


class Foo:
    def __init__(self) -> None:
        self._first_done = threading.Event()
        self._second_done = threading.Event()

    def first(self, print_first) -> None:
        print_first()
        self._first_done.set()

    def second(self, print_second) -> None:
        self._first_done.wait()
        print_second()
        self._second_done.set()

    def third(self, print_third) -> None:
        self._second_done.wait()
        print_third()


def main() -> None:
    foo = Foo()

    output_lock = threading.Lock()
    result = []

    def safe_append(value: str) -> None:
        with output_lock:
            result.append(value)

    threads = [
        threading.Thread(target=foo.third, args=(lambda: safe_append("third"),)),
        threading.Thread(target=foo.second, args=(lambda: safe_append("second"),)),
        threading.Thread(target=foo.first, args=(lambda: safe_append("first"),)),
    ]

    for thread in threads:
        thread.start()

    for thread in threads:
        thread.join()

    print("".join(result))


if __name__ == "__main__":
    main()
